Add timeouts to HealthChecks and retry checks

commonlisp opened the pull request:
Fixes issue #3532. Added timeouts for HTTP and TCP checks
and enabled HealthCheck to check health#maxRetries times
before declaring Unhealthy.

thockin commented on pkg/kubelet/util.go:
```diff
@@ -50,7 +49,7 @@ func InitHealthChecking(k *Kubelet) {
 	// in each checker's init(), iterate those here.
 	health.AddHealthChecker(health.NewExecHealthChecker(k))
-	health.AddHealthChecker(health.NewHTTPHealthChecker(&http.Client{}))
+	health.AddHealthChecker(health.NewHTTPHealthChecker(health.Timeout))
```
If we are supporting timeouts, why not on exec, too?  Set up an alarm signal before executing...

commonlisp replied:
Is there a preferred way of doing this by a utils function (it wasn't obvious from a skim of Wait)? If not, I've set up the typical select over time.After and a channel returning the regular Exec results.

thockin commented on pkg/health/health.go:
```diff
@@ -34,9 +35,12 @@ const (
 	Unknown
 )
 
+const Timeout = 5 * time.Minute
```
DefaultTimeout

thockin replied:
But shouldn't this come thru the API?

commonlisp replied:
Yes, I was considering that. I added a timeoutSeconds property in LivenessProbe. Hopefully this isn't too invasive.

commonlisp replied:
Is there an established system for falling back to defaults for API properties?

thockin commented on pkg/health/health_test.go:
```diff
@@ -131,7 +131,7 @@ func TestMuxHealthChecker(t *testing.T) {
 		container.LivenessProbe.HTTPGet.Port = util.NewIntOrStringFromString(port)
 		container.LivenessProbe.HTTPGet.Host = host
-		health, err := mc.HealthCheck("test", "", api.PodStatus{}, container)
+		health, err := mc.HealthCheck("test", "", api.PodStatus{}, container, Timeout)
```
Shouldn't we test these with a small timeout, and ensure that it fails?

commonlisp replied:
Yes, I added a test for the http health checker. Perhaps the tcp and exec checkers would also benefit from tests there.

thockin commented on pkg/health/exec.go:
```diff
@@ -43,15 +44,30 @@ func (e *ExecHealthChecker) HealthCheck(podFullName string, podUID types.UID, st
+	select {
+	case statusResult := (<-ch):
+		return statusResult.st, statusResult.err
+	case <-time.After(timeout(container)):
```
I predict a leak - a health-check command that locks up and times out, followed by a container restart, and repeat.  Is there no way to do a subprocess that we can reliably kill?  Can we fork() in go, and set an alarm signal?

thockin replied:
If we exec out to 'sh -c "alarm $timeout docker exec ..." maybe?

thockin commented on pkg/health/http_test.go:
```diff
@@ -134,4 +132,14 @@ func TestHTTPHealthChecker(t *testing.T) {
+			TimeoutSeconds: 1,
+		},
+	}
+	health, _ := hc.HealthCheck("test", "", api.PodStatus{PodIP: "0.0.0.0"}, container)
```
How do we know this exercises the timeout path?  Should you instead have a second test case that has a HandlerFunc() that just sleeps for 5 seconds?

thockin commented on pkg/health/tcp.go:
```diff
@@ -80,7 +81,7 @@ func (t *TCPHealthChecker) HealthCheck(podFullName string, podUID types.UID, sta
 		return Unknown, err
 	}
-	return DoTCPCheck(net.JoinHostPort(host, strconv.Itoa(port)))
+	return DoTCPCheck(net.JoinHostPort(host, strconv.Itoa(port)), timeout(container))
```
No test coverage for this?

smarterclayton commented on pkg/api/types.go:
```diff
@@ -283,6 +283,8 @@ type LivenessProbe struct {
 	// Length of time before health checking is activated.  In seconds.
 	InitialDelaySeconds int64 `json:"initialDelaySeconds,omitempty"`
+	// Length of time before health checking times out.  In seconds.
+	TimeoutSecs int64 `json:"timeoutSecs,omitempty"`
```
Will we ever want sub second timeouts?  I can think of some health checks that 100ms is considered failure - since it's an int64, why not make this Timeout in milliseconds (or use resource.Quantity)?

commonlisp replied:
Thanks! Yeah, I did consider that. Sub-second is probably the better choice. It is just that http.client Timeouts are in seconds. Perhaps we should do our own timeout mechanism here.

thockin replied:
I'm not sure that this sort of API is is appropriate for 100ms timeouts.  Or rather we should not do a bunch of engineering to make that work, at least not right now.

bgrant0607 replied:
We're not going to probe, kill, and restart a container in less than a second. That could lead to all sorts of problems. Seconds seems fine.

Seconds should be spelled out.

thockin commented:
This PR looks mostly good.  Maybe we want to start without exec timeouts, and handle that in a followup?  We don;t need to let it block othewise good work..

bgrant0607 commented on pkg/health/health.go:
```diff
@@ -34,6 +35,9 @@ const (
 	Unknown
 )
 
+const defaultTimeout = 5 * time.Minute
```
5 minutes? Really? Why so long?

commonlisp replied:
I'm dropping it down to 1 second.

bgrant0607 commented on pkg/health/health.go:
```diff
@@ -103,6 +113,14 @@ func findPortByName(container api.Container, portName string) int {
 	return -1
 }
 
+func timeout(container api.Container) time.Duration {
```
The default value cannot be initialized in this way. The value should be stated in the field descriptions in the versioned types.go files, and should be populated in the field such that if the user does a GET on the object, they'll see the value, as discussed in #1502 and #2587. Pending the "right" solution, the value can be initialized in pkg/api/validation/validation.go, as in #3587.

commonlisp replied:
Ah, thanks, I was looking for that.

bgrant0607 commented:
Note that PR #3695 will conflict with this change.
/cc @mikedanese

mikedanese commented:
A rebase will be nasty as pkg/health is deleted in that PR.  These changes will most likely have to be ported to pkg/probe by someone which shouldn't be too hard. Not sure if there's a better way to go about this but if this get's merged first, I can make sure these changes make it over.

bgrant0607 commented:
#3695 will go in soon.

bgrant0607 commented:
I'd also like #3818 to go in before this does.

thockin commented:
#3695 is in.  #3818 needs a rebase and an e2e run.

On Wed, Jan 28, 2015 at 11:00 AM, Brian Grant <notifications@github.com>
wrote:

> I'd also like #3818
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3818> to go in
> before this does.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3593#issuecomment-71893628>
> .
>

bgrant0607 commented:
#3818 is in.

commonlisp committed:
Add timeouts to HealthChecks and retry checks

Fixes issue #3532. Added timeouts for HTTP and TCP checks
and enabled kubelet/probe to kubelet#maxRetries times
before declaring Failure.

Added Probe.TimeoutSecs to API

Probe variants now check container.LivenessProbe.TimeoutSeconds
Also added a test for timeouts in http_test.go.

commonlisp commented:
This change has been updated for the health -> probe refactor.

bgrant0607 commented:
Thanks! API looks good.

@mikedanese Would you be willing to do the code review?

mikedanese commented on pkg/api/helpers.go:
```diff
@@ -73,3 +74,11 @@ var standardResources = util.NewStringSet(
 	return standardResources.Has(str)
 }
+
+func ProbeTimeout(container Container) time.Duration {
```
This should accept a `api.Probe` and live in `pkg/probe/probe.go` until defaults are centralized. Two reasons: We are trying to decouple probes from api objects like Container, and `container.LivenessProbe.TimeoutSeconds` looks like it could through a nil pointer dereference panic if this function is used incorrectly. Optionally, `1 * time.Second` could live in a nicely named const.

thockin replied:
Why take a probe?  Why not just take a Duration and check for 0.

This function is of dubious value, though - by the time kubelet gets it, the api-version-specific default should have been set.  @yujuhong was just refactoring around defaults.

mikedanese replied:
I agree. This is only used in one spot and should probably just be inlined until the defaults work is in, then we can just remove it.

mikedanese commented on pkg/probe/http/http.go:
```diff
@@ -28,16 +29,14 @@ import (
-func (pr *HTTPProber) Probe(host string, port int, path string) (probe.Status, error) {
-	return DoHTTPProbe(formatURL(host, port, path), pr.client)
+func (pr *HTTPProber) Probe(host string, port int, path string, timeout time.Duration) (probe.Status, error) {
+	return DoHTTPProbe(formatURL(host, port, path), &http.Client{Timeout: timeout})
```
Is there any way this wouldn't have to create a client on every probe? They are somewhat expensive. Maybe, DoHTTPProbe could accept a timeout like DoTCPProbe, and then do a manual select After timeout in there.

thockin replied:
+1

mikedanese commented:
LGTM, besides the comments which are mostly optional.

mikedanese commented on pkg/probe/http/http.go:
```diff
@@ -28,15 +29,18 @@ import (
 // Probe returns a ProbeRunner capable of running an http check.
-func (pr *HTTPProber) Probe(host string, port int, path string) (probe.Status, error) {
+func (pr *HTTPProber) Probe(host string, port int, path string, timeout time.Duration) (probe.Status, error) {
+	pr.setTimeout(timeout)
```
This is no longer thread safe as the timeout can be modified by multiple callers of Probe. It's probably better to manually implement the timeout. This isn't tested and is missing some err checks but something like:
```go
func DoHTTPProbe(url string, client HTTPGetInterface, timeout time.Duration) (probe.Status, error) {
	var resp http.Response
	req, err := http.NewRequest("GET", url, nil)
	defer req.Close()
	done := make(chan struct{}, 1)
	defer close(done)
	go func() {
		resp, err = client.Do(req)
		done <- struct{}{}
	}()
	select {
	case <-done:
		break
	case <-time.After(timeout):
		return probe.Failure, fmt.Errof("Probe timed out waiting for responseresponse timed out:")
	}
	if err != nil {
		glog.V(1).Infof("HTTP probe error: %v", err)
...
```
It's not pretty but it's probably what we want in this situation since stdlib http/client doesn't have a GetTimeout() :(

commonlisp replied:
Is client.Do guaranteed to return? I suppose the default redirect policy would likely guarantee that. If not, doesn't that potentially result in a goroutine leak similar to the prior Exec timeout handling?

mikedanese replied:
Hmm, I was thinking that req.Close() was something but it's not. You'd need to close the underlying tcp connection which should cause client.Do() to return with an error. Rereading the doc, http.Transport.CancelRequest() should work but to get one form the an http.Client requires a type assertion. Instead of `defer req.Close()` maybe:
```
if t, ok := client.Transport.(*http.Transport); ok {
	defer t.CancelRequest(req)
}
```
which is also not pretty and requires some interface changing HTTPGetInterface but shouldn't leak tcp connections or goroutines. At that point you don't even need an http.Client though, just an http.Transport. That's just one option though, there may be a better.

mikedanese replied:
@commonlisp I looked over the net/http package last night in depth. It looks like http.Transport is the main reason http.Client is so expense (as it's caching tcp connections under the hood). If we had one http.Transport per probe runner, and created a client with `http.Client{Transport: runner.transport, Timeout: timeout}` we get the ease of use of the client. It also seems that if no transport is specified when a client is created, the http.DefaultTransport is used, so creating new http.Client{} isn't as expensive as I had previously thought. I think how it was before is reasonable, or if you want each HTTPProbeRunner to have it's own http.Transport. Sorry to make you jump through hoops. What do you think?

commonlisp replied:
Sounds good! I'll look into it tonight.

mikedanese commented:
LGTM, ping @thockin

brendandburns commented:
LGTM, merging.

thockin commented:
LGTM

raindev commented:
What's the reason for specifying timeouts in seconds not milliseconds? I imagine, services working under higher load would benefit from shorter than a second health check timeouts.
